src/mds/query.py

```diff
@@ -85,6 +94,401 @@ def add_filter(query):
+
+        f = f(target_json_key).alias()
+
+        return exists(
+            db.select("*")
+            .select_from(f)
+            .where(
+                operators[scalar_operator_name]["sql_comparator"](
+                    column(f.name), scalar_right_operand
+                )
+            )
+        )
+
+    def get_all_sql_clause(target_json_key, scalar_operator_name, scalar_right_operand):
+        """
+        Generate a SQLAlchemy clause that corresponds to the :all operation.
+
+        Args:
+            target_json_key (str): the json key pointing to the array which we
+            are filtering based on
+
+            scalar_operator_name (str): the name of the operator (leading colon
+            included) (see operators dict for a list of valid operators)
+
+            scalar_right_operand (str, bool, int, float): the value that each
+            element in the target_json_key array is compared against
+
+        Returns:
+            SQLAlchemy clause corresponding to the :all operation.
+        """
+        if (
+            "type" in operators[scalar_operator_name]
+            and operators[scalar_operator_name]["type"] == "text"
+        ):
+            f = db.func.jsonb_array_elements_text
+        else:
+            f = db.func.jsonb_array_elements
+            scalar_right_operand = cast(scalar_right_operand, JSONB)
+
+        f = f(target_json_key).alias()
+        count_f = db.func.jsonb_array_length(target_json_key)
+
+        return ColumnOperators.__eq__(
+            count_f,
+            db.select([db.func.count()])
+            .select_from(f)
+            .where(
+                operators[scalar_operator_name]["sql_comparator"](
+                    column(f.name), scalar_right_operand
+                )
+            ),
+        )
+
+    operators = {
+        ":eq": {"sql_comparator": ColumnOperators.__eq__},
+        ":ne": {"sql_comparator": ColumnOperators.__ne__},
+        ":gt": {"sql_comparator": ColumnOperators.__gt__},
+        ":gte": {"sql_comparator": ColumnOperators.__ge__},
+        ":lt": {"sql_comparator": ColumnOperators.__lt__},
+        ":lte": {"sql_comparator": ColumnOperators.__le__},
+        ":like": {
+            "type": "text",
+            "sql_comparator": ColumnOperators.like,
+        },
+        ":all": {"sql_clause": get_all_sql_clause},
+        ":any": {"sql_clause": get_any_sql_clause},
+    }
+
+    def parse_filter(filter_string):
+        """
+        Parse filter_string and return an abstract syntax tree representation.
+
+        Args:
+            filter_string (str): the filter string to parse
+            (e.g. '(_uploader_id,:eq,"57")' )
+
+        Returns:
+            A parsimonious abstract syntax tree representation of
+            filter_string.
+        """
+        if not filter_string:
+            return
+
+        #  json_value and below was taken from
+        #  https://github.com/erikrose/parsimonious/pull/23/files
+        grammar = Grammar(
```

Comment:
I'd think parsing the grammar might be expensive. Can we do this once on app init and maintain a singleton `grammar` somewhere? Then just call `grammar.parse()` here?

Reply:
I should say `compiling` above rather than `parsing`. I think the `Grammar` constructor would be re-parsing the definition every time we call `parse_filter` https://github.com/erikrose/parsimonious/blob/master/parsimonious/grammar.py#L68

Reply:
Really nice catch!